support/metrics/metricsTracker.go

```diff
@@ -0,0 +1,165 @@
+package metrics
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/stellar/kelp/support/logger"
+
+	"github.com/google/uuid"
+)
+
+var (
+	amplitudeAPIURL string = "https://api2.amplitude.com/2/httpapi"
+	amplitudeAPIKey string = os.Getenv("AMPLITUDE_API_KEY")
+)
+
+// Tracker wraps the properties for Amplitude events,
+// and can be used to directly send events to the
+// Amplitude HTTP API.
+type Tracker struct {
+	client *http.Client
+	props  props
+	start  time.Time
+}
+
+type event struct {
+	UserID    string      `json:"user_id"`
+	DeviceID  string      `json:"device_id"`
+	EventType string      `json:"event_type"`
+	Props     interface{} `json:"event_properties"`
+}
+
+// props holds the properties that we need for all Amplitude events.
+// This lives on the `Tracker` struct, although
+// TODO: Add geodata.
+// TODO: Add cloud server information.
+type props struct {
+	BuildVersion string    `json:"build_version"`
+	Os           string    `json:"os"`
+	Gui          bool      `json:"gui"`
+	Strategy     string    `json:"strategy"`
+	UpdateTime   float64   `json:"update_time"`
+	Exchange     string    `json:"exchange"`
+	Pair         string    `json:"pair"`
+	SessionID    uuid.UUID `json:"session_id"`
+}
+
+// updateProps holds the properties for the update Amplitude event.
+type updateProps struct {
+	props
+	SecondsSinceStart float64 `json:"seconds_since_start"`
+}
+
+// deleteProps holds the properties for the delete Amplitude event.
+type deleteProps struct {
+	props
```

Comment:
Right - the thinking here was that the specific properties structs would embed a struct that contains the "common" fields between all of them. Embedding different structs could create a more complex sense of dependency. This also allows greater modularity, in case we decide to add information to `updateProps` that we don't want in `deleteProps`.

That being said, the above involves a good deal of future proofing, and I'm happy to make this change now. Just flagging this thinking for future changes.

Reply:
ah I see, that makes sense and I agree with your thinking and that's how I prefer to do it as well.
I was confused because we had used `props` in the initial event so thought the dependency grew out from there (so if we ever wanted to update the props sent to the initial event it would be sent everywhere else too which is bad).
Anyhow, with the suggestion to use `SecondsSinceStart` everywhere and rename to `commonProps` or similar would help disambiguate this. It would also be helpful to add a comment next to the struct definition with the thinking you described for posterity.